cardano-ledger-test/src/Test/Cardano/Ledger/Properties.hs

```diff
@@ -0,0 +1,519 @@
+
+genRecipientsFrom :: [TxOut A] -> GenRS [TxOut A]
+genRecipientsFrom txOuts = do
+  let outCount = length txOuts
+  approxCount <- lift $ choose (1, outCount)
+  let extra = outCount - approxCount
+      avgExtra = ceiling (toInteger extra % toInteger approxCount)
+      genExtra e
+        | e <= 0 || avgExtra == 0 = pure 0
+        | otherwise = lift $ chooseInt (0, avgExtra)
+  let goNew _ [] !rs = pure rs
+      goNew e (tx : txs) !rs = do
+        leftToAdd <- genExtra e
+        goExtra (e - leftToAdd) leftToAdd (inject (Coin 0)) tx txs rs
+      goExtra _ _ s tx [] !rs = genWithChange s tx rs
+      goExtra e 0 s tx txs !rs = goNew e txs =<< genWithChange s tx rs
+      goExtra e n s (TxOut _ v _) (tx : txs) !rs = goExtra e (n - 1) (s <+> v) tx txs rs
+      genWithChange s (TxOut addr v d) rs = do
+        c <- Coin <$> lift (choose (1, unCoin $ coin v))
+        r <- genTxOut (s <+> inject c)
+        if c < coin v
+          then
+            let change = TxOut addr (v <-> inject c) d
+             in pure (r : change : rs)
+          else pure (r : rs)
+  goNew extra txOuts []
+
+genValidatedTx :: GenRS (UTxO A, ValidatedTx A)
+genValidatedTx = do
+  GenEnv {geValidityInterval, gePParams} <- ask
+  UTxO utxoNoCollateral <- genUTxO
+  -- 1. Produce utxos that will be spent
+  n <- lift $ choose (1, length utxoNoCollateral)
+  toSpendNoCollateral <-
+    Map.fromList . take n <$> lift (shuffle $ Map.toList utxoNoCollateral)
+  -- 2. Check if all Plutus scripts are valid
+  let allValid vs = IsValid $ getAll $ mconcat [All v | SJust (IsValid v) <- vs]
+      toSpendNoCollateralTxOuts = Map.elems toSpendNoCollateral
+      -- We use maxBound to ensure the serializaed size overestimation
+      maxCoin = Coin (toInteger (maxBound :: Int))
+      txAddrsWithDatums = [addr | TxOut addr _ (SJust _) <- toSpendNoCollateralTxOuts]
+  isValid <- allValid <$> mapM getScriptValidity txAddrsWithDatums
+  -- 3. Generate all recipients and witnesses needed for spending Plutus scripts
+  recipients <- genRecipientsFrom toSpendNoCollateralTxOuts
+  exUnits <- genExUnits (length txAddrsWithDatums)
+  redeemerWitsList <-
+    zipWithM (uncurry genTxPlutusWitness)
+      [ (ix, dh)
+        | (ix, (_, TxOut _ _ (SJust dh))) <- zip [0 ..] (Map.toAscList toSpendNoCollateral)
+      ]
+      exUnits
+  keyWitsMakers <- mapM genTxWitness toSpendNoCollateralTxOuts
+  -- 4. Estimate inputs that will be used as collateral
+  maxCollateralCount <-
+    lift $ chooseInt (1, fromIntegral (_maxCollateralInputs gePParams))
+  bogusCollateralTxId <- lift arbitrary
+  let bogusCollateralTxIns =
+        Set.fromList
+          [ TxIn bogusCollateralTxId (fromIntegral i)
+            | i <- [maxBound, maxBound - 1 .. maxBound - maxCollateralCount - 1]
+          ]
+  collateralAddresses <-
+    replicateM maxCollateralCount $
+      genRecipientWithPayment . coerceKeyRole . KeyHashObj =<< genCredential
+  bogusCollateralKeyWitsMakers <-
```

Comment:
you are just filling out the field so that the cost in bytes is over-estimated, right?

Reply:
yep